lib/Dialect/FIRRTL/Transforms/RandomizeRegisterInit.cpp

```diff
@@ -0,0 +1,67 @@
+//===- RandomizeRegisterInit.cpp - Randomize register initialization ------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines the RandomizeRegisterInit pass.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "circt/Dialect/FIRRTL/FIRRTLOps.h"
+#include "circt/Dialect/FIRRTL/FIRRTLTypes.h"
+#include "circt/Dialect/FIRRTL/Passes.h"
+#include "mlir/IR/Builders.h"
+
+using namespace mlir;
+using namespace circt;
+using namespace firrtl;
+
+namespace {
+struct RandomizeRegisterInitPass
+    : public RandomizeRegisterInitBase<RandomizeRegisterInitPass> {
+  void runOnOperation() override;
+};
+
+} // end anonymous namespace
+
+std::unique_ptr<mlir::Pass> circt::firrtl::createRandomizeRegisterInitPass() {
+  return std::make_unique<RandomizeRegisterInitPass>();
+}
+
+void RandomizeRegisterInitPass::runOnOperation() {
+  OpBuilder builder(getOperation());
+
+  // Collect all registers.
+  SmallVector<Operation *> regs;
+  getOperation().walk([&](Operation *op) {
+    if (isa<RegOp, RegResetOp>(op))
+      regs.push_back(op);
+  });
```

Comment:
Nit: instead of storing the registers in a vector, the body of the loop over the vector of registers could be made a lambda that is called on each `RegOp` or `RegResetOp`. Minor simplification that should avoid heap allocation in the vector for modules with a lot of registers.